pkg/aws/aws.go

```diff
@@ -0,0 +1,132 @@
+package aws
+
+import (
+	"github.com/aws/aws-sdk-go/service/ec2/ec2iface"
+	"github.com/submariner-io/cloud-prepare/pkg/api"
+
+	"k8s.io/client-go/rest"
+)
+
+const (
+	messageRetrieveVPCID  = "Retrieving VPC ID"
+	messageRetrievedVPCID = "Retrieved VPC ID %s"
+)
+
+type awsCloud struct {
+	client         ec2iface.EC2API
+	k8sConfig      *rest.Config
+	gwInstanceType string
+	infraID        string
+	region         string
+}
+
+// NewCloud creates a new api.Cloud instance which can prepare AWS for Submariner to be deployed on it
+func NewCloud(k8sConfig *rest.Config, client ec2iface.EC2API, infraID, region, gwInstanceType string) api.Cloud {
```

Comment:
For ACM, we deploy gateway with `manifestwork` from the hub (broker) cluster, we don't have managed cluster kubeconfig

could we consider to provide a callback function instead of giving a kubeconfig directly, e.g. we may define a function
```
type DeployGatewayFunc func(machineSet *unstructured.Unstructured) error
type DeleteGatewayFunc func(machineSet *unstructured.Unstructured) error
```
then in `deployGateway` function, we callback it to create gateway
```
func (ac *awsCloud) deployGateway(vpcID, gatewaySecurityGroup string, publicSubnet *ec2.Subnet) error {
	amiID, err := ac.findAMIID(vpcID)
	if err != nil {
		return err
	}

	machineSet, err := ac.initMachineSet(vpcID, gatewaySecurityGroup, amiID, publicSubnet)
	if err != nil {
		return err
	}


	return c.deployGatewayFunc(machineSet)
}
```

thus we allow different invoker to implement its logic with its context

Reply:
How about we could have interfaces for gw and port setup separately and a factory mode such as:

```go

type Gateway interface {
    SetupGateway() error
    CleanGateway() error
}

type Port interface {
    SetupPort(input PrepareForSubmarinerInput) error
    CleanupPort(input PrepareForSubmarinerInput) error
}

factory.WithGatewayImpl(gateway).WithPortImpl(port).WithReporter(reporter).ToCloud()
```

So in subctl case, we could have awsCloud to implement both Gateway and Port interface. And in submariner-addon case we could have a reimplementation of gateway interface?

Reply:
we could for aws, but in some cases (e.g. vsphere on premise) we cannot access the spoke's apiserver. That is the reason we have a common way  to use manifeswork to deploy.

Reply:
yeah, the manifest work is required for us, we cannot deploy machine sets directly

The submariner-addon is running on the hub cluster, The kubeConfig [1] is the hub cluster kubeconfig, we cannot use it to deploy machine sets on managed cluster.

For our process of deploying the gateway, we define the machine sets in the manifest work and add the manifest work to managed cluster namespace on the hub cluster,  then the managed clusters pull the manifest work from hub and apply the machine sets from the manifest work,  so for submariner-addon, it is running on the hub cluster, it does not have the managed cluster hub kubeconfig

Reply:
> yeah, the manifest work is required for us, we cannot deploy machine sets directly
> 
> The submariner-addon is running on the hub cluster, The kubeConfig [1] is the hub cluster kubeconfig, we cannot use it to deploy machine sets on managed cluster.
> 
> For our process of deploying the gateway, we define the machine sets in the manifest work and add the manifest work to managed cluster namespace on the hub cluster, then the managed clusters pull the manifest work from hub and apply the machine sets from the manifest work, so for submariner-addon, it is running on the hub cluster, it does not have the managed cluster hub kubeconfig

OK thanks for the detailed explanation, I'll make the necessary adjustments

Reply:
> How about we could have interfaces for gw and port setup separately and a factory mode such as:
> 
> ```go
> type Gateway interface {
>     SetupGateway() error
>     CleanGateway() error
> }

Please see the latest commit and let me know if it looks good to you to use.

> 
> type Port interface {
>     SetupPort(input PrepareForSubmarinerInput) error
>     CleanupPort(input PrepareForSubmarinerInput) error
> }

For port setup I'm not sure why'd you need a separate functionality?
It seems that my implementation here pretty much does the same port setup that you are doing and the input allows to easily differentiate between the internal and external ports to open.
Let me know if I misunderstand.

>